### DISCUSSION TITLE
Custom Loader Plugin

### DISCUSSION THREAD

**Discussion by greg-hellings** [tox-dev/tox#2882] [Q&A] [answered]
I'm working on upgrading [my plugin](https://github.com/tox-dev/tox-ansible/) to the new tox4 method of doing things. I've cut version 1.x of the plugin to work with tox 3, so I'm ready to completely replace the hooks to the new way.

My plugin has two features that it adds:

1. A set of CLI options that will filter the active envlist
2. Auto-discovery of Ansible test cases. Each of these causes the plugin to define a Tox Environment with a set collection of requirements and test commands.

As I understand it, adding a few CLI options is still straightforward. I see several places in the current code where that happens and I believe I can update that code easily.

1. However, I don't see anywhere in the current plugin spec where I can register the code to define the Tox environments nor to filter the active envlist. Speaking on Discord I've been told that I need to instantiate a memory loader to create the environments and I think I've figured that part out. But I can't see where to define the hooks. Is there a hook that I'm missing where I can define the code that will allow me to return the instances of MemoryLoader? And is there a hook where I can tag the active ones (or is that a flag in the MemoryLoader that I define?)

Does anyone else have a plugin that's compatible with tox 4.x which does something similar that I can see to use as a crib sheet?

**Comment by gaborbernat** [comment, marked as answer]
I'm not aware of similar plugins but the advice from discord is the way to go 🙂 Here's a quick demonstration via an inline plugin seeded via `toxfile.py`.

```
❯ lsd --tree .
 .
├──  tox.ini
└──  toxfile.py
```

```python
# toxfile.py
from tox.config.loader.memory import MemoryLoader
from tox.config.sets import CoreConfigSet, EnvConfigSet
from tox.config.types import EnvList
from tox.plugin import impl
from tox.session.state import State


@impl
def tox_add_core_config(core_conf: CoreConfigSet, state: State) -> None:
    # override env list
    core_conf.loaders.insert(0, MemoryLoader(env_list=EnvList(["a", "b"])))
	# extend env list
	env_list = cast(EnvList, core_conf["env_list"])
    env_list.envs.extend(["a", "b"])

@impl
def tox_add_env_config(env_conf: EnvConfigSet, state: State) -> None:
    env_conf.loaders.insert(0, MemoryLoader(description=f"env for {env_conf.name}"))
```

```ini
# tox.ini
[tox]
env_list =
    C
```

```bash
❯ tox l
default environments:
C -> env for C
a -> env for a
b -> env for b
```

**Reply by greg-hellings** [reply]
I figured the advice is sound. But my question still remains - which hook allows me to latch onto that step in the process? I don't see a read_env_list hook nor a filter_env_list hook or anything that matches. Does this need to be an enhancement to tox first, or am I just not understanding the hooks I am seeing?

**Reply by gaborbernat** [reply]
The later I believe 🙂

**Reply by greg-hellings** [reply]
Can you help enlighten me, then?

**Reply by gaborbernat** [reply]
Edited.
